pkg/controller/certificate.go

```diff
@@ -0,0 +1,173 @@
+	r.log.V(4).Info("reconciling certificate resource", "name", certificateName)
+	fetched := &certmanagerv1.Certificate{}
+	key := types.NamespacedName{
+		Name:      desired.GetName(),
+		Namespace: desired.GetNamespace(),
+	}
+	exist, err := r.Exists(r.ctx, key, fetched)
+	if err != nil {
+		return FromClientError(err, "failed to check %s certificate resource already exists", certificateName)
+	}
+
+	if exist && recon {
+		r.eventRecorder.Eventf(es, corev1.EventTypeWarning, "ResourceAlreadyExists", "%s certificate resource already exists, maybe from previous installation", certificateName)
+	}
+	if exist && hasObjectChanged(desired, fetched) {
+		r.log.V(1).Info("certificate has been modified, updating to desired state", "name", certificateName)
+		if err := r.UpdateWithRetry(r.ctx, desired); err != nil {
+			return FromClientError(err, "failed to update %s certificate resource", certificateName)
+		}
+		r.eventRecorder.Eventf(es, corev1.EventTypeNormal, "Reconciled", "certificate resource %s reconciled back to desired state", certificateName)
+	} else {
+		r.log.V(4).Info("certificate resource already exists and is in expected state", "name", certificateName)
+	}
+	if !exist {
+		if err := r.Create(r.ctx, desired); err != nil {
+			return FromClientError(err, "failed to create %s certificate resource", certificateName)
+		}
+		r.eventRecorder.Eventf(es, corev1.EventTypeNormal, "Reconciled", "certificate resource %s created", certificateName)
+	}
+
+	return nil
+}
+
+func (r *ExternalSecretsReconciler) getCertificateObject(es *operatorv1alpha1.ExternalSecrets, resourceLabels map[string]string, fileName string) (*certmanagerv1.Certificate, error) {
+	certificate := decodeCertificateObjBytes(assets.MustAsset(fileName))
+
+	updateNamespace(certificate, es)
+	updateResourceLabels(certificate, resourceLabels)
+
+	if err := r.updateCertificateParams(es, certificate); err != nil {
+		return nil, NewIrrecoverableError(err, "failed to update certificate resource for %s/%s deployment", es.GetNamespace(), es.GetName())
+	}
+
+	return certificate, nil
+}
+
+func (r *ExternalSecretsReconciler) updateCertificateParams(es *operatorv1alpha1.ExternalSecrets, certificate *certmanagerv1.Certificate) error {
+	certManageConfig := es.Spec.ExternalSecretsConfig.WebhookConfig.CertManagerConfig
+
+	externalSecretsNamespace := certificate.GetNamespace()
+
+	dns := []string{
+		serviceExternalSecretWebhookName,
+		fmt.Sprintf("%s.%s", serviceExternalSecretWebhookName, externalSecretsNamespace),
+		fmt.Sprintf("%s.%s.svc", serviceExternalSecretWebhookName, externalSecretsNamespace),
+	}
+	certificate.Spec.DNSNames = dns
+
+	if certManageConfig.CertificateRenewBefore != nil {
+		certificate.Spec.RenewBefore = certManageConfig.CertificateRenewBefore
+	}
+
+	if certManageConfig.CertificateDuration != nil {
+		certificate.Spec.Duration = certManageConfig.CertificateDuration
+	}
+
+	if certManageConfig.IssuerRef != (operatorv1alpha1.ObjectReference{}) {
```

Comment:
```suggestion
	if certManageConfig.IssuerRef.Name == "" {
	     return fmt.Errorf("issuerRef failure log")
	}
```

And I think `assertIssuerRefExists()` should be called first in this func. Does not change much, just avoids doing other stuff which wouldn't be of use, if issuerRef doesn't exist.

Reply:
Moved the code to the start of the function.